src/midi/midi.cpp

```diff
@@ -88,43 +88,49 @@ uint8_t MIDI_message_len_by_status[256] = {
 #include "midi_alsa.h"
 #endif
 
-static std::list<std::unique_ptr<MidiHandler>> handlers = {};
+// Pointer to the static handlers list instance
+using handlers_list_t = std::list<std::unique_ptr<MidiHandler>>;
```

Comment:
Minor, but is there a reason for `std::list` over `std::vector`?  I've seen it in other places in the code as well.  I get it if you don't want to mess with this in this PR but usually vector is faster and more memory efficient unless there's a good reason not to use it.

Reply:
Yes, I would recommend defaulting to vector and only deviating from that if there's a very good reason. Linked lists are very slow because contiguousness is not guaranteed and that trumps the need reallocate once in a while by _far_ (by a very very large margin). Realloc and copying memory is very fast and a non-issue, while non-local random memory access is a huge performance killer on modern processors.

I'm not even sure what a good use case for lists are these days. Maybe huge linked lists where you must do a lot of re-linking and re-ordering all the time, i.e. when your algorithm is really designed around unlinking/relinking, etc. But even in that case vectors under a few thousand elements (or maybe even a lot more) probably wouldn't even improve anything performance-wise.

But for non-performance intensive code I would definitely choose a list or dequeue or something similar if that makes the code more _readable_, just to make this clear 😄

Reply:
>  I'm not even sure what a good use case for lists are these days.

I've used linked lists for implementing a queue (in C where I don't have the STL) because removing from the front is worst-case for a resizable array.  Even then, there's probably a better way (circular buffer or something?) because, like you said non-contiguous memory is a huge bottleneck because of cache misses.  But if you're using STL, you just have `std::queue` and `std::deque`.

I always do a double-take whenever I see `std::list` because there's usually a better data structure.  I don't think this is performance intensive code so, minor but I also don't think there's a difference in readability between `std::vector` and `std::list`.

Reply:
To put it simply: random memory access is crap for utilising your cache effectively. That's an instant performance killer, like you also mentioned.

Whereas reallocating and recopying a vector is only done once in a while, and it's a _fast_ operations (going through memory linearly both at source and destination).

Net effect: the copying on realloc is negligible compared to paying the penalty of cache misses on _every single_ operation.

Reply:
Yes, violent agreement: start with `vector` and only change when profiling proves it's a real bottleneck.

The only two cases I can pop off the top of my head are some algorithm that has incredibly frequent updates in the middle of the list along with infrequent reads, or very large elements in the list that are expensive to move.

Similar with sorting. I see people arguing over quick sort or shell sort or heap sort... for an array with 35 elements.

Reply:
Yup, some of these `std::list`s have been the result of refactoring the older hand-rolled structs that held "pointer-to-next". and it was up to the calling code to manage them (usually in non-trivial ways):

![2023-08-10_09-41](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/8edd3a63-fb08-4c6b-9051-3ed061b1b258)

Prior refactorings have migrated that to a std::list of MIDI handlers as `unique_ptr`s, where we now register and deregister the handlers, and have a get_handler:

![2023-08-10_09-51](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/b19dac5e-88a9-4fdd-9ec2-599f6fe7af20)

On a given system there are only about four handlers, so the list is very small :)

But looking at this specifically, using a `std::map` would let us lookup the handler directly by `name`, giving a perfect excuse to eliminate another std::list!

The mapper would be a good place to replace the list's with vectors beacuse that code does walk the list looking for matches on every event.. making vector a perfect fit.

Reply:
Will set this to resolved - but fully agree that a list-to-vector search 'n replace PR would be 100% welcome!

Reply:
>  Will set this to resolved - but fully agree that a list-to-vector search 'n replace PR would be 100% welcome!

Sounds good.  I think I'll grep for `std::list` later and have a look.  This code probably doesn't matter at all for performance since the list is so small but maybe there's some wins to be had elsewhere in the code.